### PR TITLE
Avoid hitting the UI thread when retrieving MEF specific services

### LINKED ISSUE #11201: All ServiceLocator calls that retrieve MEF services should avoid the UI thread.
Avoid using methods that access the UI thread when request services exposed via MEF. Use free threaded calls, or ideally uses MEF composition.

This means adding a new public method in ServiceLocator

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11201

Regression? Last working version:

## Description

Avoid hitting the UI thread when retrieving MEF specific services - V2!

This is a follow up to https://github.com/NuGet/NuGet.Client/commit/38f679c77284d79fe63b01ce034a227b72822d01, which was later reverted in https://github.com/NuGet/NuGet.Client/commit/fcb4b9a7fb52777ed3e740cda029f78813a988d9, because it had a regression. 

The regression was the following:
* Start VS
* Create an ASP.NET .NET Framework project.
* PowershellHost throws a null ref.
The issue is because in the ServiceLocator.GetGlobalServiceFreeThreadedAsync, I had *removed* the fallback.
The PackageServiceProvider is set when the NuGetPackage is initialized, and it seems like in the tested scenario it wasn't initialized on time, similar to https://github.com/NuGet/NuGet.Client/commit/fd6d1c07dd6fd96c3f2200b3b34212677addcedc. 

The change that *should* fix everything is https://github.com/NuGet/NuGet.Client/pull/4244/commits/a7a1f8d3f34b3c46fb0f3660401b5e1997666a2f.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by nkolev92** [comment]
cc @kartheekp-ms @zivkan @aortiz-msft

**Review by zivkan — APPROVED** [review]
Given that none of our existing tests caught the regression, I'm assuming that this only happens if the ASP.NET project is created as the first project/solution in a VS session, but does not happen after VS/NuGet is already fully initialized. Is that right?

If so, is it worthwhile creating an issue to track writing such an Apex test, so that we minimize the risk of similar regressions in the future? Although it does seem very implementation specific (if MEF initialization happens in a certain way, and the first solution in a VS session creates a new project that uses packages.config and the template, not the user, installs a package with tools/install.ps1....).

Although, the fallback to the global service provider makes me confused how this is related to MEF at all, or even the first/second solution in a VS session. However, if that was true, I'm surprised that none of our tests caught the bug, because I know we have tests that create ASP.NET projects.

Basically, are we going to do anything to try to prevent this type of regression again in the future?

**Comment by nkolev92** [comment]
> I'm assuming that this only happens if the ASP.NET project is created as the first project/solution in a VS session, but does not happen after VS/NuGet is already fully initialized. Is that right?

Yes, that is correct.

> If so, is it worthwhile creating an issue to track writing such an Apex test,

Yes, not sure if it'll catch this issue, but yes, created a tracking issue: https://github.com/NuGet/Client.Engineering/issues/1125. Note that machine performance likely is a factor too, so it could be one of those things that E2E machines don't catch, but RPS does.

> (if MEF initialization happens in a certain way, and the first solution in a VS session creates a new project that uses packages.config and the template, not the user, installs a package with tools/install.ps1....).

It's not about MEF initialization, it's about getting the MEF component model service from the VS Service Provider.
So it's about the NuGetPackage being initialized.
Note that we already had that fallback in the GetGlobalServiceProvider, but I had removed it in the earlier version of this change, as it effectively violated the free threadedness of the service.

> Although, the fallback to the global service provider makes me confused how this is related to MEF at all, or even the first/second solution in a VS session. However, if that was true, I'm surprised that none of our tests caught the bug, because I know we have tests that create ASP.NET projects.

We do, but those tests run on a fully loaded VS, which means the NuGetPackage is initialized.

**Comment by nkolev92** [comment]
cc @kartheekp-ms 

Feel free to revert this version if it's sitll causing RPS issues. 

I'm confident that won't be the case.
